### PR TITLE
Add support for optionally reading and writing Parquet files

### PR DESCRIPTION
**Pull request by bmcdonald3**
This PR adds the ability to read and write int64 Parquet files in Arkouda through a `pdarray.save_parquet()` function and an `ak.read_parquet()` function. These functions behave in much the same way as the `pdarray.save()` and `ak.load()` functions (see docstrings). This is only the initial support, and more dtypes will likely be added as use necessitates.

For reading Parquet files, the Arrow C++ library was chosen due to library completeness and portability. The actual functionality is written in C++, with light C wrappers to allow leveraging of the Chapel C interoperability features, which requires the C++ code to be compiled into an object file prior to calling the functions within Chapel.

On the Chapel side, the functionality is modeled off of the HDF5 code, but currently only supports int64 and int32 Arrow datatypes, which are read into int64 pdarrays.

Performance numbers collected on a single node Cray CS:
Function  | Parquet (GiB/s) | HDF5 (GiB/s)
-- | -- | --
Read | 0.65 | 2.29
Write | 0.11 | 2.81

Potential next steps:
- Better error handling
- Improve performance
- Extend to support more reading of Arrow types to pdarrays
- Extend to support Arkouda Strings
- Support append mode
- Auto detect arrow/parquet support

Parquet as an optional dependency:
Running a `make` with this PR will build Arkouda as normal, without requiring Arrow. Only when the environment variable `ARKOUDA_SERVER_PARQUET_SUPPORT` is set will the Arrow-requiring files be pulled into the build. This functions in much the same way as ZMQ and HDF5, where the path can be specified in `Makefile.paths` and there is a provided `make install-arrow` command to download Arrow from offline and build from source.

Closes https://github.com/Bears-R-Us/arkouda/issues/903

### CODE REVIEW

File: src/ArrowFunctions.cpp
Comment (glitch) [inline src/ArrowFunctions.cpp:20]: Are readers & file handles reusable or do you have to construct new ones each time?

Starting from the filename is good because you can't really screw it up, but I wonder if we need to be worried about the overhead (probably not since we generally are going to read a file once and be done).

Also another question, I noticed the reader  & file objects don't get explicitly closed.  Does that happen automatically / is it something we need to worry about?

File: src/GenSymIO.chpl
Comment (glitch) [inline src/GenSymIO.chpl:650]: 2 quick notes:

1. I'm not sure if we should throw an error here or return a MsgTuple with the `MsgType.Error` 
```chapel
return new MsgTuple(msg, MsgType.Error);
```
I know the server loop in `arkouda_sever` is in a try/catch and will handle it, so maybe it's fine.  It seems we have two ways of doing this and I'm not sure which one is "_most correct_" (note: you return a MsgType.Error tuple further below).  I'm curious what others think here.

2. OPTIONAL : I read this message a bit weird because of the passive negative voice (not sure what you call it).  Something like  "Arkouda was not built with Parquet support" might be clearer, your call.

File: src/GenSymIO.chpl
Comment (glitch) [inline src/GenSymIO.chpl:711]: Is it possible for filelist.size to ever equal zero?  We test for 1 here (and then check to see if it is blank) or else we assume it is greater than 1.

File: src/GenSymIO.chpl
Comment (glitch) [inline src/GenSymIO.chpl:738]: Minor: you have a double semi-colon here.

File: src/GenSymIO.chpl
Comment (glitch) [inline src/GenSymIO.chpl:800]: Should we return an error tuple if the type is not supported?  I know it's probably already stopped from being sent by the client.

File: src/GenSymIO.chpl
Comment (glitch) [inline src/GenSymIO.chpl:1407]: Same comment about `MsgType.Error` and error tuple as in `readAllParquetMsg`

File: src/Parquet.chpl
Comment (glitch) [inline src/Parquet.chpl:26]: If this fails it takes out the server process.  Should we throw an error here instead?

File: src/Parquet.chpl
Comment (glitch) [inline src/Parquet.chpl:45]: Yes, I agree, we should throw here.

**Review by glitch — COMMENTED** [review]
Overall this looks really good!

My main questions are about closing parquet files & readers, whether we should throw an error vs. return MsgTuple with MsgType.Error from the read/write parquet chapel functions, and the `try!` versus throwing an error.

Great work!

**Comment by ronawho** [comment]
@glitch already found this, but FYI this is based on https://github.com/Bears-R-Us/arkouda/pull/966 just because it reduces the amount of code we need to bring in for the parquet dependency check.

Ben also mentioned this a bit in the PR, but the intent of this is to be the first usable draft that we will then build on. The goal was to get something usable in that we can start getting feedback on as well as start running nightly correctness/performance testing. There are a number of known next steps (and we'd be interested in input on prioritization on those.) Note that better error-handling is one of those next steps. Today there are some halts and it's worth noting in the C++ code we currently throw an uncaught exception, which is effectively the same as a halt. Turning that into proper error-handling should be straight forward, but is a fair amount of boilerplate code and we didn't want to make those changes just yet in case larger changes to the core code were required.

File: src/ArrowFunctions.cpp
Comment (bmcdonald3) [reply src/ArrowFunctions.cpp:20]: The issue here is we can't pass a C++ object to C/Chapel, so there is no way for us to get the opened file back to Chapel to reuse.

Yes, the files are closed automaticall (see https://arrow.apache.org/docs/cpp/api/formats.html#parquet-reader)

File: src/GenSymIO.chpl
Comment (bmcdonald3) [reply src/GenSymIO.chpl:650]: Ya, I'd be open to either returning the error message or returning, so whatever anyone has a preference for I think both ways are fine. As Elliot mentioned though, we are aware the error handling is not perfect with this initial PR and were hoping to get this initial work merged before over-emphasizing the error handling (unless that is a problem).

File: src/GenSymIO.chpl
Comment (bmcdonald3) [reply src/GenSymIO.chpl:800]: Regarding this and the following error-related comments, I was anticipating this is a follow-up PR as Elliot mentioned, but I do agree there are many places errors could be handled better.

File: src/GenSymIO.chpl
Comment (bmcdonald3) [reply src/GenSymIO.chpl:711]: I suspect not -- this code was directly copy/pasted from the HDF5 code (see line 442), so since the HDF5 code seems to work just fine, I would expect this identical code to behave in the same way

File: src/ArrowFunctions.cpp
Comment (ronawho) [reply src/ArrowFunctions.cpp:20]: FWIW I don't expect there to be a noticeable overhead from not reusing. The actual file read/write should dominate the time.

**Review by glitch — APPROVED** [review]
I'd like to get a few more reviews from other team members but overall I think it looks pretty good.

**Review by pierce314159 — APPROVED** [review]
Looks good to me! Nice Work

**Review by reuster986 — CHANGES_REQUESTED** [review]
Great work, @bmcdonald3 ! I like the design here, especially how you handled the optional inclusion of parquet functionality at compile time in the makefile. I think it's a good prototype for how we can think about extending arkouda with other modular content.

I'm on board with the intention expressed by you and @ronawho to get basic functionality working in this PR and then come back and add more robust error handling, etc. later.

I just requested one minor change regarding how wildcards are handled, because I want us to adhere to `ak.read_all` usage, rather than `ak.load_all`. 

Well done!

**Comment by ronawho** [comment]
> I'm on board with the intention expressed by you and @ronawho to get basic functionality working in this PR and then come back and add more robust error handling, etc. later.

@reuster986 do you (or others) have any input on prioritization of next steps? (there's an unordered list of potential next steps in the PR description)

**Comment by bmcdonald3** [comment]
Thanks @reuster986, just pushed up some code that addresses your suggestion, assuming that I understand sufficiently. Also, I would note that the method for optional inclusion was work from @ronawho, so I cannot take credit for that!

File: tests/parquet_test.py
Comment (reuster986) [inline tests/parquet_test.py:11]: I think this should be `if (arr1 != arr2).any():`. 

As written, arrays will compare equal unless *all* of their values differ.

**Review by reuster986 — CHANGES_REQUESTED** [review]
Other than a small change on the test logic, everything looks good.

**Comment by mhmerrill** [comment]
@reuster986 @bmcdonald3 we would like to merge this week, is this ready to go or do we need to do more?

File: tests/parquet_test.py
Comment (reuster986) [reply tests/parquet_test.py:11]: @glitch I believe `arr1` and `arr2` are arkouda arrays. Although you could do `np.array_equal(arr1.to_ndarray(), arr2.to_ndarray())`, I guess I have a personal preference for staying in arkouda to avoid array transfers (even though those are fast now and these arrays are small, so at this point it's mostly a stylistic choice). 

Also, it's been a while since I looked, but I think most of the test suite uses the `(a == b).all()` pattern, so it might be good to use that for consistency.

**Comment by reuster986** [comment]
@mhmerrill I reviewed earlier this morning and requested a small logic change in the tests (see discussion above). After that, I think it's ready to go.

File: tests/parquet_test.py
Comment (bmcdonald3) [reply tests/parquet_test.py:11]: @reuster986 you are correct, `(a == b).all()` is the pattern elsewhere, and it seems that what other tests do most often is just `assertTrue`, rather than do a `compare_values` function like I was doing, so I have switched to that format for the test to be more consistent with other tests.

**Review by reuster986 — APPROVED** [review]

**Comment by bmcdonald3** [comment]
@mhmerrill This is all that we had planned adding for this initial Parquet support effort. We do have a couple of follow up PRs nearly ready that improve the error handling, performance, and support additional Parquet types among a few other small changes, but we wanted to get a few nights worth of testing data on this initial work before adding in those other changes if that sounds good to you.

**Comment by mhmerrill** [comment]
@bmcdonald3 @ronawho should we have a separate  CI process to check this? or does the CI already check this?

**Comment by ronawho** [comment]
> @bmcdonald3 @ronawho should we have a separate CI process to check this? or does the CI already check this?

I think we discussed this offline, but it's tested by default in CI testing -- https://github.com/Bears-R-Us/arkouda/pull/967/files#diff-3ab46ee209a127470fce3c2cf106b1a1dbadbb929a4b5b13656a4bc4ce19c0b8R7

**Pull request merged**
